implement add revision

kimberlythegeek opened the pull request.

kimberlythegeek commented:
Also it should be noted that the appearance of the table needs to be fixed. If the commit message is very long, it changes the height of the row, so this should be truncated, as well as the hash (similar to what was done in the search results).

It should be a fixed width so there is no horizontal scrolling. the timestamps should be converted to dates. the repository IDs should be matched to repository names. I've filed #110 for this and will add my comments there

julienw reviewed (changes requested):
Thanks, it's very good to see how the project moves forward!

I'm marking this PR as "request changes" for 2 main reasons:
* a performance reason due to the fact that all list items will be rerendered for each new check
* an architecture reason because I think some code should be moved to reducers instead of hooks

But happy to discuss about this!

eng-esther commented on src/reducers/SelectedRevisions.tsx:
```diff
@@ -48,6 +13,9 @@ const selectedRevisions = createSlice({
     },
+    setSelectedRevisions(state, action: PayloadAction<Revision[]>) {
+      state.revisions = action.payload;
+    },
```
we can try, but I think this is fine the way it is.

kimberlythegeek commented on src/hooks/useSelectRevision.ts:
```diff
@@ -0,0 +1,49 @@
+        console.log('maximum four revisions');
+      }
+    });
+    dispatch(setSelectedRevisions(newSelected));
+    dispatch(clearCheckedRevisions());
```
I would not be able to dispatch `setAlert` if I moved this code to the reducer

eng-esther commented on src/types/state.ts:
```diff
@@ -42,6 +42,12 @@ export type SearchState = {
+// contains the indices of currently checked revisions
+// in searchResults state
+export type CheckedRevisionsState = {
+  revisions: number[];
```
should this be 
```suggestion
  revisions: Revision[];
```

kimberlythegeek replied:
This state holds the index of a revision in searchResults. I can rename this to revisionIndices to make this more clear.

eng-esther commented on src/tests/SelectedRevisionsTable.test.tsx:
```diff
@@ -29,4 +60,68 @@ describe('Search View', () => {
   });
+
+  it('should print error message if trying to add more than four revisions', async () => {
+    const { testData } = getTestData();
+    global.fetch = jest.fn(() =>
+      Promise.resolve({
+        json: () => ({
+          results: testData,
+        }),
+      }),
+    ) as jest.Mock;
+    // set delay to null to prevent test time-out due to useFakeTimers
+    const user = userEvent.setup({ delay: null });
```
[nitpick] I noticed  this block occurs in 3 it blocks.
would it make sense to have a before each shared by all three?

julienw replied:
The testing library folks advise againse using the `beforeEach` thingie because they obscure what a test is doing.

However it might make sense to extract them to a common function named for example `setup` with common setup code.

kimberlythegeek commented:
@julienw 
> I feel bad that nearly everything I suggested didn't work out in the end, sorry about that. Thanks a ton for trying all of them.

no worries! I appreciate the feedback and enjoy learning

> Some problems I found while testing, but can be handled in follow-ups: 1. ![image](https://user-images.githubusercontent.com/454175/172676678-e26d9147-fc3a-4bb1-a2e5-6698c802c3ca.png)
> 
> Probably a z-index to adjust?
> 
>     2. Also we don't see the warning once we're scrolled down (this happen when 4 revisions are already added for example). Maybe we need it as `position: fixed`. (that might fix the first issue too).

I've filed #139 to fix this UI, and also found that MaterialUI does have a timed alert component called SnackBar, which I will include in that patch

>     3. I see "4" in the "Project" column after I added revisions, I'm not sure what this means.

Currently we only show the repository_id, but I will be addressing this in #110

kimberlythegeek commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -3,23 +3,29 @@ import ListItem from '@mui/material/ListItem';
 function SearchResultsListItem(props: SearchResultsListItemProps) {
-  const { checked, item, handleToggle } = props;
+  const { index, item, handleToggle } = props;
+  const isChecked: boolean = useSelector((state: RootState) =>
+    state.checkedRevisions.revisions.includes(index),
+  );
```
#140
